src/scanner.js

```diff
@@ -0,0 +1,69 @@
+import taskcluster from 'taskcluster-client'
+import scopeUtils from 'taskcluster-lib-scopes'
+import User from './user'
+var debug = require('debug')('scanner');
+
+export default async function scanner(cfg, authorizers) {
+  // * get the set of identityProviderIds
+  // * for each:
+  //   * fetch all clients, sort by identity
+  //   * for each identity:
+  //     * get roles from providers, expand
+  //     * for each client in that identity:
+  //       * get, verify client.expandedScopes satisfied by identity's expandedScopes
+
+  // NOTE: this function performs once auth operation at a time.  It is better
+  // for scans to take longer than for the auth service to be overloaded.
+  let auth = new taskcluster.Auth({credentials: cfg.app.credentials});
+
+  // gather all identityProviderIds used in any authorizer
+  let identityProviders = new Set(
+      authorizers.map(authz => authz.identityProviders).reduce((a,b) => a.concat(b)))
+
+  // enumerate all clients for any of those identity providers
+  let clients = [];
+  for (let idp of identityProviders) {
+    clients = clients.concat(await auth.listClients({prefix: idp + "/"}));
+  }
+  clients.sort((a, b) => {
+    if (a.clientId < b.clientId) return -1;
+    if (a.clientId > b.clientId) return 1;
+    return 0;
+  });
+
+  // iterate through the clients, constructing a new User as necessary, comparing
+  // the client's scopes to the User's scopes and disabling where necessary.
+  let user, userScopes;
+  let idPattern = /^([^\/]*\/[^\/]*)\/.+$/
+  for (let client of clients.sort()) {
+    debug("examining client", client.clientId);
+    if (!client.clientId.match(idPattern) || client.disabled) {
```

Comment:
It would just mean that the prefix matching in `listClients` wasn't working, and that this was a clientId we should have skipped anyway.  I think this is just being overly cautious (since I assume the pattern matches a few lines later when calling `replace`).  I'll replace it with alert-operator.